### ISSUE TITLE
defending against clock set in past problem

### ISSUE THREAD

**Issue by Francois-v3** [varnish/libvmod-vsthrottle#4]
I have just experienced vsthrottle blocking a lot of IP which doesn't reach rate limit.
This was due to fact that the clock was set 2 days before ...
Reading the source code, i guess this leads bucket tokens became a large negative number.
Could you add below defending code in calc_tokens in src/vmod_vsthrottle.c ?

```
if (b->tokens < 0)
    b->tokens = 0;
```

**Comment by daghf** [comment]
Thanks!

I've pushed 99e2a16be860718ea6a698145afb9ec3339f59fc that should take care of it.

**Comment by Francois-v3** [comment]
Not sure to post in the rigth place...

i tried this fix. It was ok on preprod but on production, with about 800 request/s i've got one assert each 15-30 seconds:

```
varnishd[23893]: Child (23272) Panic message:#012Assert error in calc_tokens(), vmod_vsthrottle.c line 92:#012 Condition(delta >= 0) not true.# ...
```

Reading the source code, i think that b-lastused can be update since you get now.

Moving down
`double now = get_ts_mono();`
juste after
`AZ(pthread_mutex_lock(&v->mtx));` 
in vmod_is_denied, fixed the issue for me in production.

Regards.

**Comment by daghf** [comment]
Thanks for the feedback - this is certainly a race condition.

I've commited your suggested fix in 8441901.
